### ISSUE TITLE
Phase 3: Deep Learning Model Development (LSTM)

### ISSUE THREAD

**Issue by dhar174** [dhar174/steel_defect_demo#6]
## **Note to Assistant: if the sub-issues are all completed, simply check over each individual sub-issue thoroughly for blocking/breaking errors, check over all tasks in the parent issue, and just overall perform final polish, ensuring compatibility with all other parts of the app (where interactions between components exists). We need a holistic, finished product. Also note: No large/major refactors if it isnt to fix app-breaking issues!**

# Phase 3: Deep Learning Model Development (LSTM)

## Overview
Implement a Long Short-Term Memory (LSTM) neural network model for sequence-based steel casting defect prediction. This phase focuses on leveraging temporal patterns in sensor data to improve upon the baseline model's performance through deep learning techniques.

## Objectives
- Develop an LSTM neural network architecture optimized for time series defect prediction
- Implement comprehensive sequence data preprocessing and normalization pipeline
- Create a robust training framework with early stopping and hyperparameter optimization
- Establish model comparison framework between baseline and deep learning approaches
- Achieve improved temporal pattern recognition over baseline XGBoost model

### 3.1 LSTM Model Architecture Implementation
- **File**: `src/models/lstm_model.py`
- **Dependencies**: torch, numpy
- **Key Features**:
  - Bidirectional LSTM layers with configurable hidden size and depth
  - Dropout regularization for overfitting prevention
  - Fully connected classifier head with sigmoid activation
  - Flexible architecture supporting different sequence lengths
  - Batch normalization and layer normalization options

### 3.2 Sequence Feature Extraction Pipeline
- **File**: `src/features/feature_extractor.py`
- **Dependencies**: pandas, numpy, scikit-learn
- **Functionality**:
  - Time series sequence extraction with sliding windows
  - Z-score normalization and standardization
  - Sequence padding and truncation for fixed-length inputs
  - Data augmentation techniques (noise injection, time warping)
  - Train/validation/test sequence splitting with temporal awareness

### 3.3 PyTorch Dataset and DataLoader Implementation
- **File**: `src/data/sequence_dataset.py`
- **Dependencies**: torch, numpy
- **Features**:
  - Custom PyTorch Dataset class for casting sequences
  - Efficient batch loading with configurable batch sizes
  - Memory-efficient data loading for large datasets
  - Support for weighted sampling to handle class imbalance
  - Data augmentation integration during training

### 3.4 LSTM Training Pipeline
- **File**: `src/models/model_trainer.py` (extended)
- **Dependencies**: torch, numpy, yaml
- **Key Features**:
  - Adam optimizer with configurable learning rate and weight decay
  - Weighted Binary Cross-Entropy loss for class imbalance handling
  - Early stopping with patience-based validation monitoring
  - Learning rate scheduling and gradient clipping
  - Comprehensive training metrics tracking and logging

### 3.5 Model Training Script
- **File**: `scripts/train_lstm_model.py`
- **Functionality**:
  - Command-line interface with configuration management
  - GPU acceleration support with automatic device detection
  - Training progress visualization and real-time metrics
  - Model checkpointing and best model saving
  - Hyperparameter logging and experiment tracking

### 3.6 Model Development Notebook
- **File**: `notebooks/03_model_development.ipynb`
- **Analysis Components**:
  - LSTM architecture exploration and hyperparameter analysis
  - Training curve visualization and convergence analysis
  - Sequence length sensitivity analysis
  - Attention visualization for model interpretability
  - Baseline vs LSTM performance comparison

### LSTM Architecture Configuration
```yaml
lstm_model:
  architecture:
    input_size: 5  # Number of sensors
    hidden_size: 64
    num_layers: 2
    dropout: 0.2
    bidirectional: false
    
  training:
    batch_size: 32
    learning_rate: 0.001
    num_epochs: 100
    early_stopping_patience: 15
    weight_decay: 1e-4
    
  data_processing:
    sequence_length: 300  # 5 minutes at 1Hz
    normalization: "z_score"
    padding: "zero"
    
  loss_function:
    type: "weighted_bce"
    pos_weight: 3.0  # For class imbalance
```

### Performance Targets
- **AUC-ROC**: > 0.88 on test set (improvement over baseline)
- **AUC-PR**: > 0.75 on test set
- **Training Time**: < 30 minutes on GPU
- **Inference Time**: < 200ms per sequence
- **Memory Usage**: < 4GB GPU memory during training

### Sequence Processing Requirements
- **Sequence Length**: 300 time steps (5 minutes at 1Hz sampling)
- **Input Features**: 5 sensor channels
- **Batch Size**: 32 sequences per batch
- **Normalization**: Z-score standardization per sensor
- **Padding Strategy**: Zero-padding for shorter sequences

## Deliverables
1. Complete `SteelDefectLSTM` model class with configurable architecture
2. `CastingSequenceDataset` PyTorch dataset implementation
3. Extended `LSTMTrainer` class with deep learning training pipeline
4. Sequence feature extraction and preprocessing pipeline
5. LSTM training script with GPU support and experiment tracking
6. Model development notebook with architecture analysis
7. Unit tests for LSTM components (`tests/test_lstm_model.py`)
8. Trained LSTM model artifacts and performance reports
9. Model comparison analysis (baseline vs LSTM)

## Testing Requirements
- Unit tests for LSTM forward pass and gradient computation
- Sequence dataset functionality and data loading tests
- Training pipeline integration tests with mock data
- Model serialization and loading functionality tests
- Performance benchmarking on different sequence lengths
- GPU/CPU compatibility testing

## Dependencies
```
torch>=1.13.0
torchvision>=0.14.0
numpy>=1.23.0
pandas>=1.5.0
scikit-learn>=1.1.0
matplotlib>=3.6.0
seaborn>=0.12.0
pyyaml>=6.0
tqdm>=4.64.0
jupyter>=1.0.0
pytest>=7.2.0
tensorboard>=2.11.0  # Optional for training visualization
```

## Execution Commands
```bash
# Train LSTM model
python scripts/train_lstm_model.py

# Train with custom configuration
python scripts/train_lstm_model.py --config configs/lstm_config_custom.yaml

# Run model development analysis
jupyter notebook notebooks/03_model_development.ipynb

# Run LSTM-specific tests
python -m pytest tests/test_lstm_model.py -v

# Compare models
python scripts/evaluate_models.py --models baseline,lstm
```

### 3.7 Attention Mechanism (Optional Enhancement)
- **File**: `src/models/attention_lstm.py`
- **Features**:
  - Self-attention layers for improved temporal modeling
  - Attention weight visualization for model interpretability
  - Multi-head attention for capturing different temporal patterns

### 3.8 Model Interpretability
- **File**: `src/models/model_explainer.py`
- **Features**:
  - SHAP values for sequence-level feature importance
  - Gradient-based saliency maps for temporal attention
  - Counterfactual analysis for defect prediction triggers

## Success Criteria
- [ ] LSTM model achieves AUC-ROC > 0.88 on test set
- [ ] Training pipeline completes successfully with early stopping
- [ ] Sequence preprocessing handles variable-length inputs correctly
- [ ] Model demonstrates improved temporal pattern recognition over baseline
- [ ] Training converges within 100 epochs with stable validation metrics
- [ ] GPU acceleration works correctly and improves training speed
- [ ] All unit tests pass with >80% code coverage
- [ ] Model artifacts are properly saved and can be loaded for inference
- [ ] Performance comparison shows clear improvement over baseline model

## Prerequisites
- [ ] Phase 1 synthetic data generation completed
- [ ] Phase 2 baseline model training completed
- [ ] PyTorch environment properly configured
- [ ] GPU support available (recommended but not required)
- [ ] Sequence data preprocessing pipeline tested

## Performance Comparison Framework
The LSTM model will be evaluated against the baseline XGBoost model using:
- Direct performance metrics comparison (AUC-ROC, AUC-PR, F1-score)
- Temporal pattern detection capabilities analysis
- Early defect warning performance (time-to-detection metrics)
- Computational efficiency comparison (training and inference time)
- Model interpretability and feature importance analysis
